File: test-scripts/test-tool-registry-refactor.kt
/**
 * Test script for the refactored ToolRegistry mechanism
 *
 * This script verifies:
 * 1. ToolRegistry respects ToolConfigService configuration
 * 2. Only enabled built-in tools are registered
 * 3. MCP tools can be dynamically registered
 * 4. CodingAgent.getAllAvailableTools() returns correct tools
 */

import cc.unitmesh.agent.config.ToolConfigFile
import cc.unitmesh.agent.config.ToolConfigService
import cc.unitmesh.agent.tool.registry.ToolRegistry
import cc.unitmesh.agent.tool.filesystem.DefaultToolFileSystem
import cc.unitmesh.agent.tool.shell.DefaultShellExecutor

fun main() {
    println("🧪 Testing ToolRegistry refactor...")
    
    // Test 1: ToolRegistry with no config (backward compatibility)
    testBackwardCompatibility()
    
    // Test 2: ToolRegistry with selective tool configuration
    testSelectiveToolConfiguration()
    
    // Test 3: ToolRegistry with empty configuration
    testEmptyConfiguration()
    
    println("✅ All tests passed!")
}

fun testBackwardCompatibility() {
    println("\n📋 Test 1: Backward compatibility (no config)")
    
    val registry = ToolRegistry(
        fileSystem = DefaultToolFileSystem(),
        shellExecutor = DefaultShellExecutor(),
        toolConfigService = null
    )
    
    val tools = registry.getAllTools()
    println("   Registered tools: ${tools.keys}")
    
    // Should register all built-in tools
    val expectedTools = setOf("read-file", "write-file", "grep", "glob", "shell")
    val actualTools = tools.keys
    
    expectedTools.forEach { toolName ->
        if (toolName !in actualTools) {
            throw AssertionError("Expected tool '$toolName' not found in registry")
        }
    }
    
    println("   ✅ All expected tools registered")
}

fun testSelectiveToolConfiguration() {
    println("\n📋 Test 2: Selective tool configuration")
    
    // Create config with only some tools enabled
    val config = ToolConfigFile(
        enabledBuiltinTools = listOf("read-file", "grep"),
        enabledMcpTools = emptyList(),
        mcpServers = emptyMap()
    )
    
    val configService = ToolConfigService(config)
    val registry = ToolRegistry(
        fileSystem = DefaultToolFileSystem(),
        shellExecutor = DefaultShellExecutor(),
        toolConfigService = configService
    )
    
    val tools = registry.getAllTools()
    println("   Registered tools: ${tools.keys}")
    
    // Should only register enabled tools
    val expectedTools = setOf("read-file", "grep")
    val actualTools = tools.keys
    
    if (actualTools != expectedTools) {
        throw AssertionError("Expected tools $expectedTools, but got $actualTools")
    }
    
    // Should NOT register disabled tools
    val disabledTools = setOf("write-file", "glob", "shell")
    disabledTools.forEach { toolName ->
        if (toolName in actualTools) {
            throw AssertionError("Disabled tool '$toolName' should not be registered")
        }
    }
    
    println("   ✅ Only enabled tools registered, disabled tools excluded")
}

fun testEmptyConfiguration() {
    println("\n📋 Test 3: Empty configuration")
    
    // Create config with no tools enabled
    val config = ToolConfigFile(
        enabledBuiltinTools = emptyList(),
        enabledMcpTools = emptyList(),
        mcpServers = emptyMap()
    )
    
    val configService = ToolConfigService(config)
    val registry = ToolRegistry(
        fileSystem = DefaultToolFileSystem(),
        shellExecutor = DefaultShellExecutor(),
        toolConfigService = configService
    )
    
    val tools = registry.getAllTools()
    println("   Registered tools: ${tools.keys}")
    
    // Should register no tools
    if (tools.isNotEmpty()) {
        throw AssertionError("Expected no tools, but got ${tools.keys}")
    }
    
    println("   ✅ No tools registered when none enabled")
}

File: test-scripts/test-tool-registry-refactor.js
/**
 * Test script for the refactored ToolRegistry mechanism
 * 
 * This script verifies:
 * 1. ToolRegistry respects ToolConfigService configuration
 * 2. Only enabled built-in tools are registered
 * 3. CodingAgent.getAllAvailableTools() returns correct tools
 */

const mod = require('../../mpp-core/build/compileSync/js/main/productionLibrary/kotlin/autodev-mpp-core.js');

const JsCodingAgent = mod.cc.unitmesh.agent.JsCodingAgent;
const JsToolConfigFile = mod.cc.unitmesh.agent.config.JsToolConfigFile;
const JsChatConfig = mod.cc.unitmesh.agent.config.JsChatConfig;
const JsKoogLLMService = mod.cc.unitmesh.llm.JsKoogLLMService;
const JsModelConfig = mod.cc.unitmesh.llm.JsModelConfig;

async function testToolRegistryRefactor() {
    console.log('🧪 Testing ToolRegistry refactor...');
    
    try {
        // Test 1: CodingAgent with selective tool configuration
        await testSelectiveToolConfiguration();
        
        // Test 2: CodingAgent with empty configuration
        await testEmptyConfiguration();
        
        // Test 3: CodingAgent with all tools enabled
        await testAllToolsEnabled();
        
        // Test 4: Backward compatibility
        await testBackwardCompatibility();
        
        console.log('✅ All tests passed!');
    } catch (error) {
        console.error('❌ Test failed:', error.message);
        process.exit(1);
    }
}

async function testSelectiveToolConfiguration() {
    console.log('\n📋 Test 1: Selective tool configuration');
    
    try {
        // Create config with only some tools enabled
        const chatConfig = new JsChatConfig(0.7, "", 4096);
        const toolConfig = new JsToolConfigFile(
            ["read-file", "grep"],  // Only enable read-file and grep
            [],  // No MCP tools
            {},  // No MCP servers
            chatConfig
        );
        
        // Create a mock LLM service with valid provider
        const modelConfig = new JsModelConfig("OPENAI", "gpt-4", "test-key", 0.7, 4096, "https://api.openai.com/v1");
        const llmService = new JsKoogLLMService(modelConfig);
        
        // Create CodingAgent with selective configuration
        const agent = new JsCodingAgent(
            "/tmp/test-project",
            llmService,
            10,  // maxIterations
            null,  // renderer
            null,  // mcpServers
            toolConfig
        );
        
        console.log('   ✅ CodingAgent created with selective tool configuration');
        console.log('   📝 Expected: Only read-file and grep tools should be available');
    } catch (error) {
        console.error('   ❌ Failed to create CodingAgent with selective config:', error.message);
        throw error;
    }
}

async function testEmptyConfiguration() {
    console.log('\n📋 Test 2: Empty configuration');
    
    try {
        // Create config with no tools enabled
        const chatConfig = new JsChatConfig(0.7, "", 4096);
        const toolConfig = new JsToolConfigFile(
            [],  // No built-in tools
            [],  // No MCP tools
            {},  // No MCP servers
            chatConfig
        );
        
        // Create a mock LLM service
        const modelConfig = new JsModelConfig("OPENAI", "gpt-4", "test-key", 0.7, 4096, "https://api.openai.com/v1");
        const llmService = new JsKoogLLMService(modelConfig);
        
        // Create CodingAgent with empty configuration
        const agent = new JsCodingAgent(
            "/tmp/test-project",
            llmService,
            10,  // maxIterations
            null,  // renderer
            null,  // mcpServers
            toolConfig
        );
        
        console.log('   ✅ CodingAgent created with empty tool configuration');
        console.log('   📝 Expected: No built-in tools should be available');
    } catch (error) {
        console.error('   ❌ Failed to create CodingAgent with empty config:', error.message);
        throw error;
    }
}

async function testAllToolsEnabled() {
    console.log('\n📋 Test 3: All tools enabled');
    
    try {
        // Create config with all tools enabled
        const chatConfig = new JsChatConfig(0.7, "", 4096);
        const toolConfig = new JsToolConfigFile(
            ["read-file", "write-file", "grep", "glob", "shell"],  // All built-in tools
            [],  // No MCP tools
            {},  // No MCP servers
            chatConfig
        );
        
        // Create a mock LLM service
        const modelConfig = new JsModelConfig("OPENAI", "gpt-4", "test-key", 0.7, 4096, "https://api.openai.com/v1");
        const llmService = new JsKoogLLMService(modelConfig);
        
        // Create CodingAgent with all tools enabled
        const agent = new JsCodingAgent(
            "/tmp/test-project",
            llmService,
            10,  // maxIterations
            null,  // renderer
            null,  // mcpServers
            toolConfig
        );
        
        console.log('   ✅ CodingAgent created with all tools enabled');
        console.log('   📝 Expected: All built-in tools should be available');
    } catch (error) {
        console.error('   ❌ Failed to create CodingAgent with all tools:', error.message);
        throw error;
    }
}

async function testBackwardCompatibility() {
    console.log('\n📋 Test 4: Backward compatibility (no config)');
    
    try {
        // Create a mock LLM service
        const modelConfig = new JsModelConfig("OPENAI", "gpt-4", "test-key", 0.7, 4096, "https://api.openai.com/v1");
        const llmService = new JsKoogLLMService(modelConfig);
        
        // Create CodingAgent without tool configuration
        const agent = new JsCodingAgent(
            "/tmp/test-project",
            llmService,
            10,  // maxIterations
            null,  // renderer
            null,  // mcpServers
            null   // No tool config - should use defaults
        );
        
        console.log('   ✅ CodingAgent created without tool configuration');
        console.log('   📝 Expected: All built-in tools should be available (backward compatibility)');
    } catch (error) {
        console.error('   ❌ Failed to create CodingAgent without config:', error.message);
        throw error;
    }
}

// Run the tests
testToolRegistryRefactor().catch(console.error);
